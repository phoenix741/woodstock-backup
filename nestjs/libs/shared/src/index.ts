export * from './authentification';
export * from './backuping';
export * from './commands';
export * from './config';
export * from './core';
export * from './files';
export * from './input-output';
export * from './maintenance';
export * from './manifest';
export * from './pool';
export * from './queue';
export * from './refcnt';
export * from './scanner';
export * from './shared';
export * from './statistics';
export * from './utils';
