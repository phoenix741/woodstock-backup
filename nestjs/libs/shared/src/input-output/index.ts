export * from './input-output.module';
export type { ProtobufMessageWithPosition } from './protobuf.interfaces';
export * from './protobuf.service';
export * from './yaml.service';
