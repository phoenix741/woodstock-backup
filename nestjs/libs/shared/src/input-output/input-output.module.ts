import { Module } from '@nestjs/common';
import { ProtobufService } from './protobuf.service';
import { YamlService } from './yaml.service';

@Module({
  providers: [ProtobufService, YamlService],
  exports: [ProtobufService, YamlService],
})
export class InputOutputModule {}
