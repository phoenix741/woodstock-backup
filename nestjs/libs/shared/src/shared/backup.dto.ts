import { Field, ObjectType } from '@nestjs/graphql';
import { Transform } from 'class-transformer';

@ObjectType()
export class Backup {
  number!: number;
  complete!: boolean;

  startDate!: number;
  endDate?: number;

  fileCount!: number;
  newFileCount!: number;
  existingFileCount!: number;

  @Transform((v) => BigInt(v.value))
  @Field(() => BigInt)
  fileSize!: bigint;
  @Transform((v) => BigInt(v.value))
  @Field(() => BigInt)
  existingFileSize!: bigint;
  @Transform((v) => BigInt(v.value))
  @Field(() => BigInt)
  newFileSize!: bigint;

  @Transform((v) => BigInt(v.value))
  @Field(() => BigInt)
  compressedFileSize!: bigint;
  @Transform((v) => BigInt(v.value))
  @Field(() => BigInt)
  existingCompressedFileSize!: bigint;
  @Transform((v) => BigInt(v.value))
  @Field(() => BigInt)
  newCompressedFileSize!: bigint;

  speed!: number;
}
