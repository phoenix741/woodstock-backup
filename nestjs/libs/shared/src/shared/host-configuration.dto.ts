import { Field, ObjectType } from '@nestjs/graphql';
import { ApiExtraModels, ApiProperty } from '@nestjs/swagger';
import { Exclude } from 'class-transformer';
import { IsNumber, Matches, Max, Min, ValidateNested } from 'class-validator';
import { Schedule } from './scheduler.dto.js';

/**
 * Part of config file.
 *
 * Store information about a share
 */
@ObjectType()
export class BackupTaskShare {
  @ApiProperty({ example: '/home' })
  name!: string;

  @ApiProperty({ example: ['/home/user'] })
  includes?: string[];

  @ApiProperty({ example: ['*.bak'] })
  excludes?: string[];
}

/**
 * Part of config file
 *
 * Store information about a DHCP Address
 */
@ObjectType()
export class DhcpAddress {
  @ApiProperty({ example: '192.168.101' })
  @Matches(/^(([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])\.){2}([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])$/)
  address!: string;

  @ApiProperty({ example: 0 })
  @Min(0)
  @Max(255)
  start!: number;

  @ApiProperty({ example: 50 })
  @Min(0)
  @Max(255)
  end!: number;
}

@ObjectType()
export class ExecuteCommandOperation {
  @ApiProperty({ example: '/bin/true' })
  command!: string;
}

@ObjectType()
export class BackupOperation {
  @ValidateNested()
  shares!: Array<BackupTaskShare>;

  @ApiProperty({ example: [] })
  includes?: Array<string>;

  @ApiProperty({ example: [] })
  excludes?: Array<string>;

  @ApiProperty({ example: 1200 })
  @IsNumber()
  timeout?: number;
}

@ApiExtraModels(BackupOperation, ExecuteCommandOperation)
@ObjectType()
export class HostConfigOperation {
  @ApiProperty({ type: [ExecuteCommandOperation] })
  @ValidateNested()
  @Field(() => [ExecuteCommandOperation])
  preCommands?: ExecuteCommandOperation[];

  @ApiProperty({ type: BackupOperation })
  @ValidateNested()
  @Field(() => BackupOperation)
  operation?: BackupOperation;

  @ApiProperty({ type: [ExecuteCommandOperation] })
  @ValidateNested()
  @Field(() => [ExecuteCommandOperation])
  postCommands?: ExecuteCommandOperation[];
}

/**
 * Config file for one Host
 *
 * Contains all information that can be used to backup a host.
 */
@ObjectType()
export class HostConfiguration {
  @ApiProperty({ type: Boolean })
  @Field(() => Boolean)
  isLocal?: boolean;

  @Exclude()
  password: string;

  @ApiProperty({ example: [] })
  addresses?: string[];

  @ValidateNested()
  dhcp?: DhcpAddress[];

  @ValidateNested()
  operations?: HostConfigOperation = new HostConfigOperation();

  @ValidateNested()
  schedule?: Schedule;
}
