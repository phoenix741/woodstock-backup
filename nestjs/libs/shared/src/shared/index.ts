export * from './backup.dto.js';
export * from './backups-files.dto.js';
export * from './host-configuration.dto.js';
export * from './scheduler.dto.js';
export * from './stats.dto.js';
export * from './tasks.dto.js';
export * from './woodstock-client.interface.js';
export * from './woodstock.constant.js';
export * from './woodstock.interface.js';
export * from './woodstock.model.js';
