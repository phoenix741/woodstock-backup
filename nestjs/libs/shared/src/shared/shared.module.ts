import { Module } from '@nestjs/common';
import { ApplicationConfigModule } from '../config';
import { LockService } from './lock.service';

@Module({
  imports: [ApplicationConfigModule],
  providers: [LockService],
  exports: [LockService],
})
export class SharedModule {}
