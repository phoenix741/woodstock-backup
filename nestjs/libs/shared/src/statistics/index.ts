export * from './stats-instant.service.js';
export * from './disk-statistics.service.js';
export * from './pool-statistics.service.js';
