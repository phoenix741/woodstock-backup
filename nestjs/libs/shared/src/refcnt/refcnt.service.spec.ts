import { Type } from '@nestjs/common';
import { Test, TestingModule } from '@nestjs/testing';
import { cp } from 'fs/promises';
import { count, from, toArray } from 'ix/asynciterable';
import { map } from 'ix/asynciterable/operators';
import { join } from 'path';
import { ProtoFileManifestJournalEntry } from '../shared/woodstock.model.js';
import { FileManifestJournalEntry } from '../shared';
import { ProtobufMessageWithPosition } from '../input-output';
import { ProtobufService } from '../input-output/protobuf.service.js';
import { PoolStatisticsService } from '../statistics';
import { ReferenceCount } from './refcnt.interface.js';
import { RefCntService } from './refcnt.service.js';

async function atomicToSnapshot(args: [string, Type<any>, AsyncIterable<unknown>, boolean]) {
  return {
    filename: args[0],
    type: args[1],
    source: await toArray(
      from(args[2]).pipe(map((v: Record<string, object>) => ({ ...v, sha256: v.sha256.toString() }))),
    ),
    append: args[3],
  };
}

describe('RefCntService', () => {
  let service: RefCntService;

  const statsService = {
    writeStatistics: jest.fn<Promise<void>, unknown[]>(),
  };

  const protobufService = {
    loadFile<T>(filename: string, type: Type): AsyncIterable<ProtobufMessageWithPosition<T>> {
      filename;
      type;
      return from([]);
    },
    async writeFile<O>(path: string, type: Type, source: AsyncIterable<O>, compress: boolean): Promise<void> {
      path;
      type;
      source;
      compress;
    },
    async atomicWriteFile<O>(path: string, type: Type, source: AsyncIterable<O>, compress: boolean): Promise<void> {
      path;
      type;
      source;
      compress;
    },
  };

  beforeEach(async () => {
    statsService.writeStatistics.mockClear();
  });

  describe('With real file', () => {
    beforeEach(async () => {
      const module: TestingModule = await Test.createTestingModule({
        providers: [ProtobufService, RefCntService, { provide: PoolStatisticsService, useValue: statsService }],
      }).compile();

      service = module.get<RefCntService>(RefCntService);
    });

    it('read a file list/journal and generate a refcount file', async () => {
      const journalFilename = join(__dirname, 'fixtures', 'home.filelist');
      const refcntFilename = join(__dirname, 'REFCNT');

      async function* it() {
        const protobufService = new ProtobufService();
        const it = protobufService.loadFile<FileManifestJournalEntry>(journalFilename, ProtoFileManifestJournalEntry);

        for await (const manifest of it) {
          for (const chunk of manifest.message?.manifest?.chunks || []) {
            yield {
              sha256: chunk,
              refCount: 1,
              size: 4_000_000,
              compressedSize: 4_000_000,
            };
          }
        }
      }

      try {
        await service.writeRefCnt(from(it()), refcntFilename);

        const source = service.readRefCnt(refcntFilename);
        const array = await toArray(source);
        expect(array.length).toBe(1_064);
      } finally {
        await service.deleteRefcnt(refcntFilename);
      }
    }, 120_000);

    it('addBackupRefcntTo', async () => {
      // GIVEN
      const refcntFilename = join(__dirname, 'fixtures', 'REFCNT');
      const compactDir = join(__dirname, 'compact');
      const cnt = new ReferenceCount(compactDir, compactDir, compactDir);
      try {
        await cp(refcntFilename, cnt.backupPath);

        expect(await count(service.readRefCnt(cnt.backupPath))).toBe(1_000);

        // WHEN
        await service.addBackupRefcntTo(cnt.poolPath, cnt.backupPath, cnt.unusedPoolPath);

        // THEN
        const source = service.readRefCnt(cnt.poolPath);
        const array = await toArray(source);
        expect(array.length).toBe(936);
        expect(array.reduce((acc, v) => acc + v.refCount, 0)).toBe(1_000);

        expect(statsService.writeStatistics).toBeCalledTimes(1);
        expect(
          (statsService.writeStatistics as jest.Mock<Promise<void>, [object[], string]>).mock.calls[0][0],
        ).toMatchSnapshot('statsService.writeStatistics');
      } finally {
        await service.deleteRefcnt(cnt.backupPath).catch(() => void 0);
        await service.deleteRefcnt(cnt.poolPath).catch(() => void 0);
      }
    }, 240_000);

    it('cleanup', async () => {
      const refcntFilename = join(__dirname, 'fixtures', 'REFCNT');
      const cleanupDir = join(__dirname, 'cleanup');
      const unusedPath = join(cleanupDir, 'UNUSED');
      const cnt = new ReferenceCount(cleanupDir, cleanupDir, cleanupDir);
      try {
        await cp(refcntFilename, cnt.backupPath);

        await service.removeBackupRefcntTo(cnt.backupPath, refcntFilename, unusedPath);

        const source = service.readRefCnt(cnt.backupPath);
        const array = await toArray(source);
        expect(array.length).toBe(0);

        const unusedSource = service.readUnused(unusedPath);
        const unusedArray = await toArray(unusedSource);
        expect(unusedArray.length).toBe(936);
      } finally {
        await service.deleteRefcnt(cnt.backupPath).catch(() => void 0);
        await service.deleteRefcnt(unusedPath).catch(() => void 0);
      }
    });

    it('cleanup some file', async () => {
      // GIVEN
      const refcntFilename = join(__dirname, 'fixtures', 'pool', 'REFCNT.pool');
      const unusedFilename = join(__dirname, 'fixtures', 'pool', 'REFCNT.unused');
      const backupFilename = join(__dirname, 'fixtures', 'pool', 'REFCNT.backup');

      const cleanupDir = join(__dirname, 'cleanup2');
      const newRefcntPool = join(cleanupDir, 'REFCNT');
      const newUnused = join(cleanupDir, 'UNUSED');

      try {
        await cp(unusedFilename, newUnused);
        await cp(refcntFilename, newRefcntPool);

        const source1 = service.readRefCnt(newRefcntPool);
        const array1 = await toArray(source1);
        expect(array1.length).toBe(518_508);

        const unusedSource1 = service.readUnused(newUnused);
        const unusedArray1 = await toArray(unusedSource1);
        expect(unusedArray1.length).toBe(27_456);

        await service.removeBackupRefcntTo(newRefcntPool, backupFilename, newUnused);

        const source = service.readRefCnt(newRefcntPool);
        const array = await toArray(source);
        expect(array.length).toBe(518_505);

        const unusedSource = service.readUnused(newUnused);
        const unusedArray = await toArray(unusedSource);
        expect(unusedArray.length).toBe(27_459);
      } finally {
        await service.deleteRefcnt(newRefcntPool).catch(() => void 0);
        await service.deleteRefcnt(newUnused).catch(() => void 0);
      }
    }, 240_000);
  });

  describe('With mock', () => {
    beforeEach(async () => {
      const module: TestingModule = await Test.createTestingModule({
        providers: [
          RefCntService,
          { provide: ProtobufService, useValue: protobufService },
          { provide: PoolStatisticsService, useValue: statsService },
        ],
      }).compile();

      service = module.get<RefCntService>(RefCntService);
    });

    it('cleanup some file', async () => {
      // GIVEN
      protobufService.loadFile = jest
        .fn()
        .mockReturnValueOnce(
          from([
            {
              message: {
                sha256: Buffer.from('sha256_1'),
                refCount: 5,
                size: 4_000_000,
                compressedSize: 4_000_000,
              },
            },
            {
              message: {
                sha256: Buffer.from('sha256_2'),
                refCount: 5,
                size: 4_000_000,
                compressedSize: 4_000_000,
              },
            },
            {
              message: {
                sha256: Buffer.from('sha256_3'),
                refCount: 5,
                size: 4_000_000,
                compressedSize: 4_000_000,
              },
            },
            {
              message: {
                sha256: Buffer.from('sha256_10'),
                refCount: 5,
                size: 4_000_000,
                compressedSize: 4_000_000,
              },
            },
          ]),
        )
        .mockReturnValueOnce(
          from([
            {
              message: {
                sha256: Buffer.from('sha256_2'),
                refCount: 3,
                size: 4_000_000,
                compressedSize: 4_000_000,
              },
            },
            {
              message: {
                sha256: Buffer.from('sha256_3'),
                refCount: 7,
                size: 4_000_000,
                compressedSize: 4_000_000,
              },
            },
            {
              message: {
                sha256: Buffer.from('sha256_10'),
                refCount: 5,
                size: 4_000_000,
                compressedSize: 4_000_000,
              },
            },
          ]),
        )
        .mockReturnValueOnce(
          from([
            {
              message: {
                sha256: Buffer.from('sha256_3'),
              },
            },
            {
              message: {
                sha256: Buffer.from('sha256_4'),
              },
            },
            {
              message: {
                sha256: Buffer.from('sha256_5'),
              },
            },
          ]),
        );
      protobufService.writeFile = jest.fn();
      protobufService.atomicWriteFile = jest.fn();

      // WHEN
      await service.removeBackupRefcntTo('fileToChange', 'backupPath', 'unusedPath');

      // THEN
      expect(protobufService.atomicWriteFile).toBeCalledTimes(2);
      expect(
        await atomicToSnapshot(
          (protobufService.atomicWriteFile as jest.Mock<Promise<void>, [string, Type, AsyncIterable<unknown>, boolean]>)
            .mock.calls[0],
        ),
      ).toMatchSnapshot('atomicWriteFileUnused');
      expect(
        await atomicToSnapshot(
          (protobufService.atomicWriteFile as jest.Mock<Promise<void>, [string, Type, AsyncIterable<unknown>, boolean]>)
            .mock.calls[1],
        ),
      ).toMatchSnapshot('atomicWriteFileRefCnt');
      expect(statsService.writeStatistics).toBeCalledTimes(1);
      expect(
        (statsService.writeStatistics as jest.Mock<Promise<void>, [object[], string]>).mock.calls[0],
      ).toMatchSnapshot('statsService.writeStatistics');
    });
  });
});
