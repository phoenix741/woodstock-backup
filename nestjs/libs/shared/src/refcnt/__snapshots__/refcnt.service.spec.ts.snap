// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RefCntService With mock cleanup some file: atomicWriteFileRefCnt 1`] = `
Object {
  "append": undefined,
  "filename": "fileToChange",
  "source": Array [
    Object {
      "compressedSize": 4000000,
      "refCount": 5,
      "sha256": "sha256_1",
      "size": 4000000,
    },
    Object {
      "compressedSize": 4000000,
      "refCount": 2,
      "sha256": "sha256_2",
      "size": 4000000,
    },
  ],
  "type": Object {
    "fields": Object {
      "compressedSize": Object {
        "id": 4,
        "type": "int32",
      },
      "refCount": Object {
        "id": 2,
        "type": "int32",
      },
      "sha256": Object {
        "id": 1,
        "type": "bytes",
      },
      "size": Object {
        "id": 3,
        "type": "int32",
      },
    },
  },
}
`;

exports[`RefCntService With mock cleanup some file: atomicWriteFileUnused 1`] = `
Object {
  "append": undefined,
  "filename": "unusedPath",
  "source": Array [
    Object {
      "sha256": "sha256_4",
    },
    Object {
      "sha256": "sha256_5",
    },
    Object {
      "sha256": "sha256_3",
    },
    Object {
      "sha256": "sha256_10",
    },
  ],
  "type": Object {
    "fields": Object {
      "sha256": Object {
        "id": 1,
        "type": "bytes",
      },
    },
  },
}
`;

exports[`RefCntService With mock cleanup some file: statsService.writeStatistics 1`] = `
Array [
  PoolStatistics {
    "compressedSize": 8000000n,
    "longestChain": 5,
    "nbChunk": 2,
    "nbRef": 7,
    "size": 8000000n,
  },
  ".",
]
`;

exports[`RefCntService With real file compact: statsService.writeStatistics 1 1`] = `
PoolStatistics {
  "compressedSize": 3744000000n,
  "longestChain": 10,
  "nbChunk": 936,
  "nbRef": 1000,
  "size": 3744000000n,
}
`;

exports[`RefCntService With real file compact: statsService.writeStatistics 2 1`] = `
PoolStatistics {
  "compressedSize": 3744000000n,
  "longestChain": 10,
  "nbChunk": 936,
  "nbRef": 1000,
  "size": 3744000000n,
}
`;

exports[`RefCntService With real file compact: statsService.writeStatistics 3 1`] = `
PoolStatistics {
  "compressedSize": 3744000000n,
  "longestChain": 10,
  "nbChunk": 936,
  "nbRef": 1000,
  "size": 3744000000n,
}
`;
